src/lib/rehype/rehypeImg.ts

```diff
@@ -40,6 +65,68 @@ const getImageSize = (src: string, dir: string) => {
   return sizeOf(src)
 }
 
+
+/**
+ * Sets image placeholders for the given array of images.
+ * 
+ * @param images - The array of images to set placeholders for.
+ * @param srcPath - The source page path for the images.
+ * @returns A promise that resolves to void.
+ */
+const setImagePlaceholders = async (images: ImageNode[], srcPath: string): Promise<void> => {
+  // Generate kebab-case filename from srcPath, ie: /content/nft => content-nft-data.json
+  const FILENAME = path.join(srcPath, "data.json").replaceAll("/", "-").slice(1)
+
+  // Make directory for current page if none exists
+  if (!fs.existsSync(PLACEHOLDER_IMAGE_DIR)) fs.mkdirSync(PLACEHOLDER_IMAGE_DIR, { recursive: true })
+
+  const DATA_PATH = path.join(PLACEHOLDER_IMAGE_DIR, FILENAME)
+  const existsCache = fs.existsSync(DATA_PATH)
+  const placeholdersCached: PlaceholderData = existsCache ? JSON.parse(fs.readFileSync(DATA_PATH, "utf8")) : {}
+  const placeholdersClone: PlaceholderData = structuredClone(placeholdersCached)
+
+  // Generate placeholder for internal images
+  for (const image of images) {
+    const { src } = image.properties
+
+    // Skip externally hosted images
+    if (src.startsWith("http")) continue
+
+    // Load image data from file system as buffer
+    const buffer: Buffer = fs.readFileSync(path.join("public", src))
+
+    // Get hash fingerprint of image data (no security implications; fast algorithm prioritized)
+    const hash = await getHashFromBuffer(buffer, { algorithm: "SHA-1", length: 8 })
+
+    // Look for cached placeholder data with matching hash
+    const cachedPlaceholder: Placeholder | null = placeholdersClone[src]?.hash === hash ? placeholdersClone[src] : null
+
+    // Get base64 from cached placeholder if available, else generate new placeholder
+    const { base64 } = cachedPlaceholder || await getPlaiceholder(buffer)
+
+    // Assign base64 placeholder data to image node `blurDataURL` property
+    image.properties.blurDataURL = base64
+    image.properties.placeholder = "blur"
+
+    // If cached value was not available, add newly generated placeholder data to clone
+    if (!cachedPlaceholder) {
+      placeholdersClone[src] = { hash, base64 }
+    }
+  }
+
+  // If placeholderClone is empty, delete DATA_PATH JSON file and return
+  if (Object.keys(placeholdersClone).length === 0) {
+    fs.rmSync(DATA_PATH, { recursive: true, force: true })
+    return
+  }
+
+  // If cached value is identical to clone, return without writing to file system
+  if (JSON.stringify(placeholdersCached) === JSON.stringify(placeholdersClone)) return
```

Comment:
(not blocking)

nitpick: if we care about performance, we could just use a boolean flag to know this, set when `cachedPlaceholder` is false.

nitpick2: if we choose to go into the previous direction, you wouldn't need to use `placeholdersClone`. You just load the existing placeholders and modify/add new ones on top.